### PR TITLE
Make dagster a PEP-561 compliant typed library

### PR DESCRIPTION
**Pull request by smackesey**
This PR aims to expose Dagster as a PEP-561-compliant typed Python library. This should significantly improve the development experience of users working with Dagster in the context of a language server like VSCode's Pyright/Pylance.

While some users are already getting typing info from Dagster, not all of them are. For PyLance (VSCode), if `useLibraryCodeForTypes` setting is on, then PyLance will try to pull inline type info even from libraries that aren't PEP-561-compliant. So users with this setting on should already be getting Dagster type info during development. However, the UX can be bad since the server doesn't know what's private vs public. This leads to auto-imports resolving to the wrong location (e.g. `from dagster.core... import` instead of `from dagster import`) and the autocomplete menu getting noisy. This is a problem with all libraries, so many users probably turn off this setting. And other Python language servers may not expose a setting like this at all.

This problem is solved by a [nascent standard](https://github.com/python/typing/blob/master/docs/source/libraries.rst) for declaring a library's public API. This standard is not in a PEP yet but has been implemented in PyRight/PyLance and documented in the official `python/typing` repo.

This PR takes steps towards conforming Dagster's public API to this standard. It does four things:

- Adds a `py.typed` marker file to `dagster` (no extension libs yet)
- Marks all `dagster` submodules private by prefixing the top-level submodules with `_` (there is unfortunately no way around this-- see discussion with Pyright maintainer linked below)
- Renames some other internal modules to avoid naming collisions with exports. This is a subtle issue, but if a submodule has the same name as an export from its parent, then you end up with an unstable referent for that name that is particularly problematic for static analysis. There were several places in the Dagster codebase where this collision was occurring, I renamed the modules accordingly:
    - All submodules of `dagster.core.definitions.decorators` (e.g. `dagster.core.decorators.solid` conflicted with the `solid` export from `dagster.core.decorators`, was renamed to `dagster.core.decorators.solid_decorator`
    - `dagster.core.definitions.reconstructable` > `dagster.core.definitions.reconstructable_definition`
    - `dagster.core.asset_defs.asset` > `dagster.core.asset_defs.assets`
- Adds module name mapping to `ConfigurableClassData` and `dagster.yaml` loading to support backcompatibility.

Doing this allows us to run `pyright --verifytypes dagster` to analyze coverage:

```
$ pyright --verifytypes dagster

Symbols exported by "dagster": 263
  With known type: 212
  With ambiguous type: 9
  With unknown type: 42
  Functions without docstring: 25
  Functions without default param: 59
  Classes without docstring: 0

Other symbols referenced but not exported by "dagster": 4751
  With known type: 3475
  With ambiguous type: 70
  With unknown type: 1206

Type completeness score: 80.6%

Completed in 3.699sec
```

So we're most of the way to being fully typed. Because we aren't for now, the `py.typed` file contains a single line `partial\n`, this should be deleted when the library reaches 100%.

Relevant background:

- [Typing Python Libraries](https://github.com/python/typing/blob/master/docs/source/libraries.rst) doc from `python/typing` repo.
- [PEP 561](https://peps.python.org/pep-0561/)
- [Discussion with Pyright maintainer](https://github.com/microsoft/pyright/issues/3147) about public API definition
- [Thread by Max](https://threads.com/34390936956) that touches on this issue from a little more than a year ago.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #7017** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/7017" target="_blank">Graphite</a>  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

**Comment by sryza** [comment]
I really like the idea of making dagster a typed library.

The idea of changing the paths of all dagster subpackages feels very scary to me. It feels like it could have a lot of unknown consequences. The known consequences I can think of:
* Breaks all outstanding PRs
* Breaks users who are depending on private APIs (which we generally discourage, but there are a number of cases where this is the best option)

If we make Dagster a typed library without adding the underscores, will we be making users' lives harder than they were pre-typed? Even if it does add a little bit of inconvenience for users, my impulse would be to start with making dagster a typed library, and then have a longer conversation about whether the sub-package renames are worth it.

Thoughts @alangenfeld ?

**Comment by smackesey** [comment]
> The idea of changing the paths of all dagster subpackages feels very scary to me

I felt the same way, but per [my discussion with the Pyright maintainer](https://github.com/microsoft/pyright/issues/3147) it does seem like this is an official Python standard now. Seems like a bullet that needs to be bit eventually.

>If we make Dagster a typed library without adding the underscores, will we be making users' lives harder than they were pre-typed? Even if it does add a little bit of inconvenience for users, my impulse would be to start with making dagster a typed library, and then have a longer conversation about whether the sub-package renames are worth it.

One big issue with just introducing `py.typed` without properly defining the public API is that we'll probably end up with a lot of users importing directly from submodules instead of the public API at the top level. That's because tools will auto-resolve imports to these modules. This is something IMO we really want to avoid.

A potential alternative that might work is to add `__all__ = []` at the bottom of all `dagster` submodules. I _think_ this will have the same effect, obviously it's uglier and harder to maintain going forward but wouldn't cause the breakages you fear.

**Comment by smackesey** [comment]
@sryza So I went ahead and tried plan B, adding `__all__ = []` to all of our submodules. Unfortunately the results aren't what I was expecting:

```
Symbols exported by "dagster": 6665
  With known type: 3262
  With ambiguous type: 234
  With unknown type: 3169
  Functions without docstring: 3134
  Functions without default param: 345
  Classes without docstring: 394

Other symbols referenced but not exported by "dagster": 5418
  With known type: 4897
  With ambiguous type: 36
  With unknown type: 485

Type completeness score: 48.9%

Completed in 6.847sec
```

Compared to the result from underscoring, this seems to be having no effect. I'm not sure whether I misunderstood the standard, it's a bug in pyright, or something else but I have an inquiry open with the maintainer. It's also possible that the only relevant difference is in the completeness metric, and the UX is basically the same between the two methods (i.e. symbols from modules with `__all__ = []` still won't show in autocomplete.

**Comment by smackesey** [comment]
OK, another possible solution is to patch `sys.meta_path`. I just pushed a commit that puts this at the beginning of the `dagster/__init__.py`, and on initial experimentation it seems to work:

```python
class SubModuleMapper(MetaPathFinder):
    """Maps imports of the form `dagster.XXX` to `dagster._XXX`. Backcompat
    layer to allow importing modules under old, non-underscored names after all
    top-level submodules were underscored to mark them private for
    type-checkers.
    """

    def find_spec(
        self, fullname: str, _path: t.Optional[t.List[str]], _target: t.Optional[ModuleType]
    ) -> t.Optional[ModuleSpec]:
        parts = fullname.split(".")
        if len(parts) >= 2 and parts[0] == "dagster" and parts[1][0] != "_":
            target_name = ".".join([parts[0], f"_{parts[1]}", *parts[2:]])
            return importlib.util.find_spec(target_name)
        else:
            return None


sys.meta_path.append(SubModuleMapper())
```

Need to experiment to find any unknown unknowns or edge cases, but this may allow us to do the module rename while dynamically resolving old-name imports, thus not breaking PRs or code using private APIs.

Here's a nice tutorial on`sys.meta_path`: https://dev.to/dangerontheranger/dependency-injection-with-import-hooks-in-python-3-5hap

**Comment by dpeng817** [comment]
Making Dagster a typed library seems like it would be super powerful. The breaking changes seem super real, however. In addition to just using private APIs, it sounds like any users who are already using autocompletion might have been pointed towards the full path instead of `from dagster import ...`, and it sounds like we would break all of those people as well. 

This could potentially be a good candidate for a major release / 1.0 in my opinion, but seems too invasive for a dot release certainly.

**Comment by smackesey** [comment]
@dpeng817 You might have missed this because Github is hiding my comments in the middle of the large chain of events, but with the `sys.meta_path` injection explained in previous comment (and currently on this branch) this does not create any _obvious_ breaking changes-- there may be edge cases, but I haven't encountered them yet.

**Comment by alangenfeld** [comment]
>  However, the UX can be bad since the server doesn't know what's private vs public. This leads to auto-imports resolving to the wrong location

Does pep-561 compliance change how pylance treats the library for auto-import / auto-complete? My assumption would be that if all we did was add `py.typed` we would go from a library with poor auto-_ UX & no type information to a library with poor auto-_ UX & some type information. The problems you describe are all true today. Is this incorrect?

I am very skeptical about the current approach in the PR of moving all top level modules to `_` prefix all in one go. It is incredibly difficult to review, and I don't think all submodules should get the same treatment (`check` stands out). 

If we are going to reorganize all these modules to make them private, it's unclear that just `_` prefixing the existing name is better than moving towards some more well thought out organization scheme. Where we stand today is largely incremental evolution without any clear guidelines. 

#7040 is novel, but it's unclear that we need something so invasive and systemic. I think there are a few symbols that we want to protect (may be entirely addressed by serdes loader changes) but largely we should feel safe to move things. Our examples, docs, and internal repo are probably the best signal on what to be worried about, but it is very difficult to discern standouts with the current approach. I would guess there are many cases that would be solved by changing to existing top level import, and then a handful of edge cases things that merit adding a new top level import. 

I am bought in on the end goal of a clear public API surface that is 100% typed, but the best path to get there is the tricky question.

**Comment by dpeng817** [comment]
> @dpeng817
> 
> > it sounds like any users who are already using autocompletion might have been pointed towards the full path instead of from dagster import ...
> 
> I think that's probably right.
> 
> But-- you might have missed this because Github is hiding my comments in the middle of the large chain of events in this PR, but with the `sys.meta_path` injection explained in previous comment (and currently stacked on this PR in #7040) those imports continue to work, and thus this doesn't create any _obvious_ breaking changes-- there may be edge cases, but I haven't encountered them yet.

@smackesey thanks for the clarification - yea unfortunately missed that in the github flow.

**Comment by smackesey** [comment]
@alangenfeld Sorry it took me a bit to get answers to these questions, had to seek clarification from `pyright` maintainer.

>Does pep-561 compliance change how pylance treats the library for auto-import / auto-complete? My assumption would be that if all we did was add py.typed we would go from a library with poor auto-_ UX & no type information to a library with poor auto-_ UX & some type information. The problems you describe are all true today. Is this incorrect?

Let's split this into two questions: (1) what will happen if we just add `py.typed`; (2) what will happen if we mark submodules private on top of the `py.typed` addition?

1. What will happen if we just add `py.typed`?

- `mypy`, `pyright`, and other type checkers will start using Dagster type info while users are checking their code. Right now, if `dagster` is installed normally in site-packages, all of its type info is ignored.
- For users of `pylance` or `pyright` who have `useLibraryCodeForTypes` set to off (default on in `pylance`, default off in `pyright`), they will begin to see completion suggestions from Dagster-- no autocomplete appears for non-py.typed libraries if this setting is off.

Based on the above, I think we should definitely add `py.typed`, regardless of what direction we go with marking submodules private. Especially since I have now confirmed that auto-imports, even without private submodules, will correctly resolve to top-level dagster.

2. What will happen if we mark submodules private on top of the py.typed addition?

- We'll be able to use `pyright --verifytypes` to evaluate coverage of our public API.
- Completion experience _might_ change now for `pylance` (it won't for `pyright`) and will almost certainly change in the future for both. Specifically, private symbols will be hidden from the completion menu, providing better UX.
- Unknown potential benefits from other tools that incorporate the new public API standard.

Unfortunately it's hard to find the exact completion rules for pyright/pylance spelled out anywhere, so these answers are based on my [back-and-forth with Eric Traut](https://github.com/microsoft/pyright/issues/3147). But it looks like the completion rules will at least eventually converge with type-checker public/private symbols rules (possibly soon).

After having researched this issue a bit more in the few days, it appears we can get a lot of immediate UX benefits from simply adding `py.typed`, so we should definitely do that. However, I still think it's important to formally delineate the public API by marking submodules private. It is not urgent, but adoption of this standard is going to grow and tools are going to rely on it more and more. I think it is a bullet we will need to bite eventually, and it's better to do it sooner rather than later. Though I'm certainly open to more incremental approaches than the current PR.

>I don't think all submodules should get the same treatment (check stands out).

Could you elaborate here? Do you think `check` should be part of Dagster's public API? In the current PR, all `_check` imports have been aliased to `check`, so it doesn't alter the many invocations of `check.XXX` throughout the codebase.

> If we are going to reorganize all these modules to make them private, it's unclear that just _ prefixing the existing name is better than moving towards some more well thought out organization scheme. Where we stand today is largely incremental evolution without any clear guidelines.

Marking stuff private vs internal reorg don't seem at odds to me-- IMO getting through an initial marking stuff private will just give us more freedom/confidence in reorganizing the internal module structure.

>#7040 is novel, but it's unclear that we need something so invasive and systemic. I think there are a few symbols that we want to protect (may be entirely addressed by serdes loader changes) but largely we should feel safe to move things.

I don't have a particularly strong opinion here. I'm fairly confident in #7040, and similar approaches appear to have been used in other big libraries sucessfully (see [quantumlib/Cirq](https://github.com/quantumlib/Cirq/pull/3917)), but agree it's kind of a hack and has unknown unknowns.

**Comment by alangenfeld** [comment]
> Could you elaborate here? Do you think check should be part of Dagster's public API?

The thinking when i wrote it was was that it was worth considering. Not sure that line of thought holds much water.

Given #7135, should this PR be pivoted in to focusing on just the package moves? Potentially abandon this and start a fresh one if desired. 

If #7040 is key to this approach, it seems that it should be stacked below, and that the shuffling PR (or PRs) should contain only `git mv`-ed files and no callsite changes built on top of that. No need for custom `serdes` logic in a world where we inject in to import machinery. 

When I open the code changes in the current state of the PR, seeing all the callsite changes is unsetting and i think doesn't effectively portray the approach you've iterated towards.

File: CHANGES.md [hunk 1]
```diff
@@ -3561,7 +3561,7 @@ compatible with 0.8.0
     as follows:
 
     ```python
-    from dagster.core.storage.event_log.migration import migrate_event_log_data
+    from dagster._core.storage.event_log.migration import migrate_event_log_data
```
Comment (alangenfeld) [inline CHANGES.md:3564]: dont rewrite history in change log

File: docs/content/deployment/dagster-instance.mdx [hunk 2]
```diff
@@ -89,22 +89,22 @@ To configure Run Storage, you should set the `run_storage` attribute in your `da
...
+<PyObject module="dagster._core.storage.runs" object="SqliteRunStorage" /> uses a
 Sqlite DB as the backing run storage.
 
 ```yaml file=/deploying/dagster_instance/dagster.yaml startafter=start_marker_run_storage_sqlite endbefore=end_marker_run_storage_sqlite
 # there are two ways to set run_storage to SqliteRunStorage
 
 # this config manually sets the directory (`base_dir`) for Sqlite to store run information in:
 run_storage:
-  module: dagster.core.storage.runs
+  module: dagster._core.storage.runs
```
Comment (alangenfeld) [inline docs/content/deployment/dagster-instance.mdx:100]: [1] whatever approach we take - the instance config shouldn't need to change

**Review by alangenfeld — CHANGES_REQUESTED** [review]
spoke offline about my previous comment, the utility of #7040 is limited since static tools don't understand it. 

 
I personally bias towards a "simple & honest" approach for handling back-compat of just leaving dangling references in the original file locations for things like the instance components. 

Trying to do all the moves in one PR is very difficult to review. I can sort of do it commit by commit but the `core` move is too big for the browser to handle and may need subdivision. Alternatively separate PRs may allow for incremental progress and to "canary" the change with some lesser submodules. The changes will need siblings for the `internal` repo as well.


There are lots of cases of [1] in the amount of code change i can load before chrome stalls/ crashes. 

To your queue to update this PR to its new focus or start a new set.

**Comment by smackesey** [comment]
>the utility of https://github.com/dagster-io/dagster/pull/7040 is limited since static tools don't understand it.

In a way the inability of static tools to understand #7040 is the point-- the motivation of _this_ PR is to clearly communicate the public API to static tools. The immediate (and only possible AFAICT) solution to this is to rename modules, which will break code at runtime that imports those modules. But #7040 gets us the definition of the public API for static tools without the runtime breakage. Any runtime breakage solution also understood by static tools defeats the purpose.

I'd additionally argue that any linter errors this PR will surface for code that imports our private APIs is actually a benefit. It will help migrate users to using the public API without breaking their code. If there are edge case submodules that are not in fact private, then we should just include those in the public API.

Since you've said you're bought in to the idea of a clear public API surface, I guess I don't understand why you see the linter warnings for private API use as a problem.  Do you agree that _eventually_ it would be good to surface linter warnings for private API use, and you just think it would be too abrupt to suddenly start "enforcing" the public API at the linter level? If so , maybe we should do some preliminary announcement where we tell users that uses of our private APIs are now "deprecated", and that they will raise linter errors in X releases? Or do you see linter warnings for private API use as inherently problematic?

>I personally bias towards a "simple & honest" approach for handling back-compat of just leaving dangling references in the original file locations for things like the instance components.

I'm not sure how many cases like this there are that you would want to leave in place, but IMO a dynamic solution in this case (like the small amount of serdes code in this PR) is superior. That's because:

- leaving dangling references makes the codebase more confusing. We would have both `dagster.core` and `dagster._core`, with `core` still public and visible to static analysis tools. Pretty confusing to users, especially those who need to migrate any direct imports from `dagster.core` to our public API: "core is now officially private, we renamed it to `_core`, but actually there's still "core" that your autocomplete will still suggest but you should just ignore it."
- How long would you want to leave this up for? Given the confusion of having both `core` and `_core`, leaving it permanently seems like a bad idea. In contrast, I don't think there's any issue with leaving the dynamic serdes solution up indefinitely.

Related question, since I don't fully grok instance configuration-- are users supposed to be referencing submodules in their instance config? If so, shouldn't we expose whatever they're supposed to be targeting through top-level `dagster` like everything else we want them to use?

>Trying to do all the moves in one PR is very difficult to review. I can sort of do it commit by commit but the core move is too big for the browser to handle and may need subdivision. Alternatively separate PRs may allow for incremental progress and to "canary" the change with some lesser submodules. The changes will need siblings for the internal repo as well.
>
>There are lots of cases of [1] in the amount of code change i can load before chrome stalls/ crashes.

I'm sympathetic here and will break the smaller submodules into separate PRs and put `core` in its own PR with multiple commits.

**Comment by alangenfeld** [comment]
> I don't understand why you see the linter warnings for private API use as a problem

I'm sorry, i didn't do a good job clearly communicating. I was only attempting to refer back to my previous comments misunderstanding about leaning on #7040 to do file moves without import updates for our codebase. 


> are users supposed to be referencing submodules in their instance config?

As it currently stands yes. The current system is cumbersome and needs an extensive overhaul.

 If so, shouldn't we expose whatever they're supposed to be targeting through top-level dagster like everything else we want them to use?

We don't expect users to import and interact with these objects in python directly, so it's not clear to me they should be top level imports. Asking users to change this yaml file content to change the import path is busy work with no upside for them so i don't think its an appealing path.

>  a dynamic solution in this case (like the small amount of serdes code in this PR) is superior

I think these are compelling arguments. I bet seeing the changes in a more focused context will also help my stupid brain.

> I don't want to proceed until we've worked out a vision

Getting buy-in from more of the core team for the move in general is probably best before proceeding. Also would be good to hear others opinions on backcompat strategy. 

> doubts about #7040

It's more fear than doubt, of unknown unknowns over time. If the plan is to delete soonish in like the `0.15.0` release that eases things.

**Comment by smackesey** [comment]
Closing in favor of incremental strategy discussed in this [Thread](https://threads.com/34423461702).

**Pull request closed without merging**